plugins/fapolicyd.c

```diff
@@ -0,0 +1,212 @@
+        rpmlog(RPMLOG_DEBUG, "Open: %s -> %s\n", state->fifo_path, strerror(errno));
+        goto bad;
+    }
+
+    if (stat(state->fifo_path, &s) == -1) {
+        rpmlog(RPMLOG_DEBUG, "Stat: %s -> %s\n", state->fifo_path, strerror(errno));
+        goto bad;
+    } else {
+        if (!S_ISFIFO(s.st_mode)) {
+            rpmlog(RPMLOG_DEBUG, "File: %s exists but it is not a pipe!\n", state->fifo_path);
+            goto bad;
+        }
+
+        /* keep only file's permition bits */
+        mode_t mode = s.st_mode & ~S_IFMT;
+
+        /* we require pipe to have 0660 permission */
+        if (mode != 0660) {
+        
+            rpmlog(RPMLOG_ERR, "File: %s has %o instead of 0660 \n",
+                    state->fifo_path,
+                    mode );
+
+            goto bad;
+        }
+    }
+
+    state->fd = fd;
+    /* considering success */
+    return RPMRC_OK;
+
+bad:
+    if (fd > 0)
+        close(fd);
+    return RPMRC_FAIL;
+}
+
+static rpmRC write_fifo(struct fapolicyd_data* state, const char * str)
+{
+    ssize_t len = strlen(str);
+    ssize_t written = 0;
+    ssize_t n = 0;
+
+    while (written < len) {
+        if ((n = write(state->fd, str + written, len - written)) < 0) {
+            if (errno == EINTR || errno == EAGAIN)
+                continue;
+            rpmlog(RPMLOG_DEBUG, "Write: %s -> %s\n", state->fifo_path, strerror(errno));
+            goto bad;
+        }
+        written += n;
+    }
+
+    return RPMRC_OK;
+
+ bad:
+    return RPMRC_FAIL;
+}
+
+static rpmRC fapolicyd_init(rpmPlugin plugin, rpmts ts)
+{
+    if (rpmtsFlags(ts) & (RPMTRANS_FLAG_TEST|RPMTRANS_FLAG_BUILD_PROBS))
+        goto end;
+
+    if (!rstreq(rpmtsRootDir(ts), "/"))
+        goto end;
+
+    (void) open_fifo(&fapolicyd_state);
+
+ end:
+    return RPMRC_OK;
+}
+
+static void fapolicyd_cleanup(rpmPlugin plugin)
+{
+    if (fapolicyd_state.fd > 0)
+        (void) close(fapolicyd_state.fd);
+
+    fapolicyd_state.fd = -1;
+}
+
+
+static rpmRC fapolicyd_psm_pre(rpmPlugin plugin, rpmte te)
+{
+    if (fapolicyd_state.fd == -1)
+        goto end;
+
+    if (rpmteType(te) == TR_ADDED)
+        fapolicyd_state.installed_package = 1;
+    else
+        fapolicyd_state.installed_package = 0;
+
+ end:
+    return RPMRC_OK;
+}
+
+static rpmRC fapolicyd_tsm_post(rpmPlugin plugin, rpmts ts, int res)
+{
+    if (rpmtsFlags(ts) & (RPMTRANS_FLAG_TEST|RPMTRANS_FLAG_BUILD_PROBS))
+        goto end;
+
+    /* we are ready */
+    if (fapolicyd_state.fd > 0) {
+        /* send a signal that transaction is over */
+        (void) write_fifo(&fapolicyd_state, "1\n");
+        /* flush cache */
+        (void) write_fifo(&fapolicyd_state, "2\n");
+    }
+
+ end:
+    return RPMRC_OK;
+}
+
+static rpmRC fapolicyd_scriptlet_pre(rpmPlugin plugin, const char *s_name,
+                                     int type)
+{
+    if (fapolicyd_state.fd > 0 &&
+        fapolicyd_state.changed_files > 0 &&
+        fapolicyd_state.installed_package ) {
```

Comment:
The if condition is on same indentation level as the following code, making it hard to see where the condition stops and code starts. There are various ways to deal with that, my favourite generally is moving (parts of) the condition into a helper variable, with well-named variables this has the effect of making the code self-documenting as well. For example:
```
    int need_flush = (fapolicyd_state.changed_files > 0) && fapolicyd_state.installed_package;
    if (fapolicyd_state.fd > 0 && need_flush) {
        ...
    }
```